File: 148-sort-list.cpp
/*
  Source : https://oj.leetcode.com/problems/sort-list/
  Author : Alex Xu
  Date   : Mar 04, 2015
  Problem:
  Sort a linked list in O(n log n) time using constant space complexity.

  Solution: MergeSort list.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /*
     * Merge Sort:
     * 1) break the current list into two lists: L and R
     * 2) sort L, and sort R
     * 3) merge L and R
     */
    ListNode *sortList(ListNode *head) {
        if (head == NULL || head->next == NULL)
            return head;

        ListNode * middle = getMiddleNode(head);
        ListNode * headR = NULL;
        if (middle != NULL) {
            headR = middle->next;
            middle->next = NULL;
        }

        ListNode *headL = sortList(head);
        headR = sortList(headR);
        head = merge(headL, headR);

        return head;
    }


    /*
      @ 21-merge-two-sorted-lists.cpp for more solutions to merge two sorted lists
    */
    ListNode *merge(ListNode *headL, ListNode *headR) {
        if (headL == NULL)
            return headR;
        if (headR == NULL)
            return headL;

        if (headL->val > headR->val)
            std::swap(headL, headR);

        ListNode *head = headL; // make headL the head
        headL = headL->next;

        ListNode *current = head;
        // link left list (headL) and right list(headR) to head
        while (true) {
            if (headL == NULL) {
                current->next = headR;
                break;
            }

            if (headR == NULL) {
                current->next = headL;
                break;
            }
            if (headL->val <= headR->val) {
                current->next = headL;
                current = headL;
                headL = headL->next;
            } else {
                current->next = headR;
                current = headR;
                headR = headR->next;
            }
        }

        return head;
    }

    ListNode *getMiddleNode(ListNode *head) {
        if (head == NULL || head->next == NULL)
            return head;
            
        ListNode *slow = head;
        ListNode *fast = head->next;

        while (fast != NULL && fast->next != NULL) {
            slow = slow->next; // step 1
            fast = fast->next->next;
        }

        return slow;
    }
};

File: 148-sort-list.py
"""
  Source : https://oj.leetcode.com/problems/sort-list/
  Author : Alex Xu
  Date   : Apr 30, 2018
  Problem:
  Sort a linked list in O(n log n) time using constant space complexity.

  Solution: MergeSort list.
  Note recursion is not using constant space, will enhance it later!
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        
        l1, l2 = self.break_list(head)
        l1 = self.sortList(l1)
        l2 = self.sortList(l2)
        return self.merge_sorted_list(l1, l2)
        
    def break_list(self, head):      
        pre, slow, fast = None, head, head
        
        while fast and fast.next:
            pre, slow, fast = slow, slow.next, fast.next.next
        pre.next = None
        
        return (head, slow)
    
    def merge_sorted_list(self, l1, l2):
        dummy = ListNode(0)
        tail = dummy
    
        while l1 and l2:
            if l1.val < l2.val:
                tail.next, tail, l1 = l1, l1, l1.next
            else:
                tail.next, tail, l2 = l2, l2, l2.next
        tail.next = l1 or l2
    
        return dummy.next
